skimage/exposure/_adapthist.py

```diff
@@ -122,74 +123,103 @@ def _clahe(image, kernel_size, clip_limit, nbins):
     minimum and maximum value as the input image. A clip limit smaller than 1
     results in standard (non-contrast limited) AHE.
     """
-    row_step = int(image.shape[0] // np.ceil(image.shape[0] / kernel_size[0]))
-    col_step = int(image.shape[1] // np.ceil(image.shape[1] / kernel_size[1]))
+    # pad the image so that the shape in each dimension
+    # is a multiple of the relevant kernel_size
+    pad_end_per_dim = [0] * image.ndim
+    for dim in range(image.ndim):
+        pad_end_per_dim[dim] = (kernel_size[dim] - image.shape[dim] % kernel_size[dim]) % kernel_size[dim]
 
-    nr = image.shape[0] // row_step
-    nc = image.shape[1] // col_step
+    image = np.pad(image,
+                   [[0, pad_end_per_dim[dim]] for dim in range(image.ndim)],
+                   mode='reflect',
+                   )
+
+    ns = [int(np.ceil(image.shape[dim] / kernel_size[dim]))
+          for dim in range(image.ndim)]
+
+    steps = [int(np.floor(image.shape[dim] / ns[dim]))
+             for dim in range(image.ndim)]
 
     bin_size = 1 + NR_OF_GRAY // nbins
     lut = np.arange(NR_OF_GRAY)
     lut //= bin_size
 
-    map_array = np.zeros((nr, nc, nbins), dtype=int)
+    map_array = np.zeros(tuple(ns) + (nbins,), dtype=int)
 
     # Calculate graylevel mappings for each contextual region
-    i0 = 0
-    for r in range(nr):
-        i1 = i0 + row_step
-        j0 = 0
-        for c in range(nc):
-            j1 = j0 + col_step
-            sub_img = image[i0:i1, j0:j1]
-
-            if clip_limit > 0.0:  # Calculate actual cliplimit
-                clim = max(int(clip_limit * sub_img.size), 1)
+    for inds in np.ndindex(*ns):
+
+        region = tuple([slice(inds[dim] * steps[dim], (inds[dim] + 1)
+                              * steps[dim])
+                        for dim in range(image.ndim)])
+        sub_img = image[region]
+
+        if clip_limit > 0.0:  # Calculate actual clip limit
+            clim = int(clip_limit * sub_img.size)
+            if clim < 1:
+                clim = 1
+        else:
+            clim = NR_OF_GRAY  # Large value, do not clip (AHE)
+
+        hist = lut[sub_img.ravel()]
+        hist = np.bincount(hist)
+        hist = np.append(hist, np.zeros(nbins - hist.size, dtype=int))
+        hist = clip_histogram(hist, clim)
+        hist = map_histogram(hist, 0, NR_OF_GRAY - 1, sub_img.size)
+        map_array[inds] = hist
+
+    # Interpolate greylevel mappings to get CLAHE image
+    offsets = [0] * image.ndim
+    lowers = [0] * image.ndim
+    uppers = [0] * image.ndim
+    starts = [0] * image.ndim
+    prev_inds = [0] * image.ndim
+
+    for inds in np.ndindex(*[ns[dim] + 1 for dim in range(image.ndim)]):
```

Comment:
So if I understand well `ns[dim]` is the number of cells / blocks for a given dimension. Do you really need to add 1 to this number? If yes please explain :-).

Reply:
The need for `n+1` is indeed not exactly straight forward: As you say, `ns[dim]` is the number of contextual regions for each dim (to calculate graylevel mappings). The number of blocks to process separately for interpolation is however different from this number in the general case, as becomes clear from [this](https://en.wikipedia.org/w/index.php?title=Adaptive_histogram_equalization&oldid=936814673#Efficient_computation_by_interpolation) wikipedia schematic (which I also linked in a code comment to clarify). Depending on how many pixels are padded at the end of the dim (to have the contextual regions well defined), `ns[dim]` blocks can be sufficient to process all pixels within the shape of the input image. Otherwise `ns[dim]+1` blocks are needed.

The 2d version in `master` always processes `ns[dim]+1` blocks, however thanks to your comment I went through it and realised the last block can in many cases be dropped (as it only processes pixels which will be discarded during unpadding. Thanks :) Let me know if you agree and what you think!